### PR TITLE
Type hierarchy for SPIR-V

### PR DESCRIPTION
**Pull request by antiagainst**
Another offspring from #219.

Sorry for this gigantic PR. Lots of the code is just dummy boilerplate.

### CODE REVIEW

**Comment by antiagainst** [comment]
This is the first step for adding a pass to unify types.

File: test/opt/test_type_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+      SpvTools(SPV_ENV_UNIVERSAL_1_1).BuildModule(text);
+  manager.AnalyzeType(*module);
+
+  ASSERT_EQ(7u, manager.NumTypes());
+  ASSERT_EQ(0u, manager.NumForwardPointers());
+  // Make sure we get ids correct.
+  ASSERT_EQ("uint32", manager.GetType(5)->str());
+  ASSERT_EQ("float32", manager.GetType(6)->str());
+
+  for (const auto id1 : {1, 2, 3, 4, 7}) {
```
Comment (Qining) [inline test/opt/test_type_manager.cpp]: I think it would be better with some blurbs

File: source/opt/module.h [hunk 2]
```diff
@@ -94,6 +94,7 @@ class Module {
   // Returns a vector of pointers to type-declaration instructions in this
   // module.
   std::vector<Instruction*> types();
+  std::vector<const Instruction*> types() const;
```
Comment (dneto0) [inline source/opt/module.h]: Needs a blurb. Also, it does non trivial work,  so it needs a captialized first letter.

File: source/opt/type_manager.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  Type* CreateType(const spvtools::ir::Instruction& inst);
+  // Attaches the decoration encoded in |inst| to a type. Does nothing if the
+  // given instruction is not a decoration instruction or not decorating a type.
+  void AttachDecoration(const spvtools::ir::Instruction& inst);
+
+  IdToTypeMap id_to_type_;  // Mapping from ids to their type representations.
+  ForwardPointerVector forward_pointers_;  // All forward pointer declarations.
+  // All unresolved forward pointer declarations.
+  // Refers the contents in the above vector.
+  std::unordered_set<ForwardPointer*> unresolved_foward_pointers_;
```
Comment (dneto0) [inline source/opt/type_manager.h]: "unresolved_forward_pointers_"  (missing r )

File: source/opt/type_manager.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  // Analyzes the types and decorations on types in the given |module|.
+  void AnalyzeType(const spvtools::ir::Module& module);
+
+ private:
+  // Creates and returns a type from the given SPIR-V |inst|. Returns nullptr if
+  // the given instruction is not for defining a type.
+  Type* CreateType(const spvtools::ir::Instruction& inst);
+  // Attaches the decoration encoded in |inst| to a type. Does nothing if the
+  // given instruction is not a decoration instruction or not decorating a type.
+  void AttachDecoration(const spvtools::ir::Instruction& inst);
```
Comment (dneto0) [inline source/opt/type_manager.h]: Prefer calling this "AttachIfTypeDecoration"  ? or "RecordIfTypeDecoration" ?

File: source/opt/type_manager.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  // Returns the number of types hold in this manager.
+  size_t NumTypes() const { return id_to_type_.size(); }
+
+  // Returns the forward pointer type at the given |index|.
+  ForwardPointer* GetForwardPointer(uint32_t index) const;
+  // Returns the number of forward pointer types hold in this manager.
+  size_t NumForwardPointers() const { return forward_pointers_.size(); }
+
+  // Analyzes the types and decorations on types in the given |module|.
+  void AnalyzeType(const spvtools::ir::Module& module);
```
Comment (dneto0) [inline source/opt/type_manager.h]: What if we call it a second time, on a different module, or on a modified version of the module?

Comment (dneto0) [inline source/opt/type_manager.h]: If you made the module a paramter to the (only) constructor, then it would be more RAII-like.

Comment (antiagainst) [inline source/opt/type_manager.h]: Good call. Doen.

File: source/opt/type_manager.h (hunk 4 above)
Comment (antiagainst) [inline source/opt/type_manager.h]: Done.

File: source/opt/type_manager.h (hunk 3 above)
Comment (antiagainst) [inline source/opt/type_manager.h]: Done.

File: source/opt/module.h (hunk 2 above)
Comment (antiagainst) [inline source/opt/module.h]: Done.

File: test/opt/test_type_manager.cpp (hunk 1 above)
Comment (antiagainst) [inline test/opt/test_type_manager.cpp]: Done.

**Comment by dneto0** [comment]
+1

**Comment by Qining** [comment]
lgtm +1

**Comment by antiagainst** [comment]
1 + 1 = 2.

I'll submit.

**Pull request closed without merging**